### PR TITLE
Add mesh upgrade commands for az aks mesh

### PR DESCRIPTION
**Pull request by deveshdama**
Add mesh upgrade commands for az aks mesh

az aks mesh get-upgrades -g rg-name -n mc-name
az aks mesh get-revisions -l location

### Related command

- az aks mesh get-upgrades -g rg-name -n mc-name
- az aks mesh get-revisions -l location
- az aks mesh upgrade start --revision mesh-revision -g rg-name -n mc-name
- az aks mesh upgrade complete -g rg-name -n mc-name
- az aks mesh upgrade rollback -g rg-name -n mc-name

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/aks-preview/HISTORY.rst [hunk 1]
```diff
@@ -12,6 +12,11 @@ To release a new version, please select a new version number (usually plus 1 to
 Pending
 +++++++
 
+0.5.158
++++++++
+* Add `get-upgrades` and `get-versions` to the `az aks mesh` command.
```
Comment (SanyaKochhar) [inline src/aks-preview/HISTORY.rst:17]:
```suggestion
* Add `get-upgrades` and `get-revisions` to the `az aks mesh` command.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 2]
```diff
@@ -2645,6 +2645,62 @@
         text: az aks mesh disable-ingress-gateway --resource-group MyResourceGroup --name MyManagedCluster --ingress-gateway-type Internal
 """
 
+helps['aks mesh get-revisions'] = """
+    type: command
+    short-summary: Discover available Azure Service Mesh revisions.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2650]:
```suggestion
    short-summary: Discover available Azure Service Mesh revisions and their compatibility.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 3]
```diff
@@ -2645,6 +2645,62 @@
...
+    long-summary: This command discovers available Azure Service Mesh revisions and their compatibility for the given location.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2651]:
```suggestion
    long-summary: This command discovers available Azure Service Mesh revisions and their compatibility information for the given location.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 4]
```diff
@@ -2645,6 +2645,62 @@
...
+    examples:
+      - name: Discover Azure Service Mesh revisions.
+        text: az aks mesh get-revisions --location westus2
+        crafted: true
+"""
+
+helps['aks mesh get-upgrades'] = """
+    type: command
+    short-summary: Discover available Azure Service Mesh upgrades.
+    long-summary: This command discovers available Azure Service Mesh upgrades for all the installed revisions on the cluster.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2661]:
```suggestion
    long-summary: This command discovers available Azure Service Mesh upgrades for the mesh revision installed on the cluster.
```

Comment (SanyaKochhar) [reply src/aks-preview/azext_aks_preview/_help.py:2661]: the command actually fails if there are multiple revisions present

File: src/aks-preview/azext_aks_preview/_help.py [hunk 5]
```diff
@@ -2645,6 +2645,62 @@
...
+    examples:
+      - name: Discover Azure Service Mesh upgrades.
+        text: az aks mesh get-upgrades --resource-group MyResourceGroup --name MyManagedCluster
+"""
+
+helps['aks mesh upgrade start'] = """
+    type: command
+    short-summary: Initiate Azure Service Mesh upgrade.
+    long-summary: This command initiates Azure Service Mesh upgrade to the specified revision.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2670]:
```suggestion
    long-summary: This command initiates upgrade of Azure Service Mesh to the specified revision.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 6]
```diff
@@ -2645,6 +2645,62 @@
...
+helps['aks mesh upgrade'] = """
+    type: group
+    short-summary: Commands to manage the upgrades for Azure Service Mesh.
+    long-summary: A group of commands to manage the upgrades for Azure Service Mesh in given cluster.
+"""
+
+helps['aks mesh upgrade complete'] = """
+    type: command
+    short-summary: Complete Azure Service Mesh upgrade.
+    long-summary: This command completes Azure Service Mesh upgrade in progress.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2689]:
```suggestion
    long-summary: This command completes Azure Service Mesh canary upgrade by removing the previous revision.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 7]
```diff
@@ -2645,6 +2645,62 @@
...
+    examples:
+      - name: Complete Azure Service Mesh upgrade.
+        text: az aks mesh upgrade complete --resource-group MyResourceGroup --name MyManagedCluster
+"""
+
+helps['aks mesh upgrade rollback'] = """
+    type: command
+    short-summary: Rollback Azure Service Mesh upgrade.
+    long-summary: This command rolls back Azure Service Mesh upgrade in progress.
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_help.py:2698]:
```suggestion
    long-summary: This command rolls back Azure Service Mesh upgrade to the previous stable revision.
```

File: src/aks-preview/azext_aks_preview/_params.py [hunk 8]
```diff
@@ -925,6 +923,12 @@ def load_arguments(self, _):
         c.argument('root_cert_object_name')
         c.argument('cert_chain_object_name')
 
+    with self.argument_context('aks mesh get-revisions') as c:
+        c.argument('location', required=True, help='Location to discover available Azure Service Mesh revisions.')
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/_params.py:927]:
```suggestion
        c.argument('location', required=True, help='Location in which to discover available Azure Service Mesh revisions.')
```

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 9]
```diff
@@ -2191,6 +2190,25 @@ def update_azure_service_mesh_profile(self) -> ServiceMeshProfile:
             new_profile.istio.certificate_authority.plugin.cert_chain_object_name = cert_chain_object_name
             updated = True
 
+        # deal with mesh upgrade commands
+        mesh_upgrade_command = self.raw_param.get("mesh_upgrade_command", None)
+        if mesh_upgrade_command is not None:
+            if len(new_profile.istio.revisions) < 2:
+                raise ArgumentUsageError('Azure Service Mesh upgrade is not in progress.')
```
Comment (SanyaKochhar) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:2197]: would this piece be called during upgrade start as well?

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 10]
```diff
@@ -7475,6 +7475,56 @@ def test_aks_azure_service_mesh_with_pluginca(self, resource_group, resource_gro
...
+        ])
+
+        # start an upgrade
+        upgrade_start_cmd = 'aks mesh upgrade start --resource-group={resource_group} --name={name} --revision asm-1-17'
+        self.cmd(upgrade_start_cmd, checks=[
+            self.check('serviceMeshProfile.istio.revisions[0]', 'asm-1-17'),
+        ])
+
+        # since we do not support any other revision other than asm-1-17 at this time, testing for the following commands is not possible
+        # az aks mesh upgrade rollover --resource-group={resource_group} --name={name}
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:7519]: a little bit curious, from above code, seems the supported command is rollback, so what's this rollover for?

**Review by FumingZhang — COMMENTED** [review]
lgtm

please resolve conflicts and check live test result

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 10 above)
Comment (deveshdama) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py]: this is a typo. Fixed the comment in the latest iteration

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 9 above)
Comment (deveshdama) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py]: No, since mesh_upgrade_command is set to None for upgrade start command

**Pull request closed without merging**